### ISSUE TITLE
ingressgateway hpa shows unknown cpu utilization

### ISSUE THREAD

**Issue by aditya-prasad** [istio/istio#15979]
**Bug description**
ingressgateway hpa unable to figure out cpu utilization and thus doesn't scale. all other hpa (pilot, policy, and telemetry) created by istio working fine.

might be related to #11350

```
kubectl -n istio-system get hpa,pods
```
```
NAME                                                       REFERENCE                         TARGETS         MINPODS   MAXPODS   REPLICAS   AGE
horizontalpodautoscaler.autoscaling/istio-ingressgateway   Deployment/istio-ingressgateway   <unknown>/70%   2         10        2          9m5s
horizontalpodautoscaler.autoscaling/istio-pilot            Deployment/istio-pilot            2%/80%          1         5         1          89d
horizontalpodautoscaler.autoscaling/istio-policy           Deployment/istio-policy           8%/80%          1         5         1          89d
horizontalpodautoscaler.autoscaling/istio-telemetry        Deployment/istio-telemetry        21%/80%         1         5         1          89d

NAME                                          READY   STATUS    RESTARTS   AGE
pod/grafana-7f4d444dd5-pcxff                  1/1     Running   0          176m
pod/istio-citadel-67d66d65d7-t9srg            1/1     Running   0          176m
pod/istio-galley-dfcc75c46-ljn27              1/1     Running   0          116m
pod/istio-ingressgateway-7497b4c87b-n6pg8     2/2     Running   0          9m7s
pod/istio-ingressgateway-7497b4c87b-thdcn     2/2     Running   0          8m50s
pod/istio-pilot-659cbcf679-p9grb              2/2     Running   0          178m
pod/istio-policy-6ccbf85f88-rzh5l             2/2     Running   0          178m
pod/istio-sidecar-injector-79f667994b-ql7p2   1/1     Running   0          176m
pod/istio-telemetry-5db544fc5d-2krjp          2/2     Running   1          116m
pod/istio-tracing-79db5954f-p8glb             1/1     Running   0          176m
pod/kiali-68677d47d7-6qqnp                    1/1     Running   0          176m
pod/prometheus-5977597c75-qq4qq               1/1     Running   0          176m
```

looks like metric-server is configured properly
```
metrics-server:
    Container ID:  docker://725f5bf24d2c9efe1d88aeffddcb36648f5ddc9c62bd5f670a60817cdf568b5d
    Image:         k8s.gcr.io/metrics-server-amd64:v0.3.1
    Image ID:      docker-pullable://k8s.gcr.io/metrics-server-amd64@sha256:78938f933822856f443e6827fe5b37d6cc2f74ae888ac8b33d06fdbe5f8c658b
    Port:          443/TCP
    Host Port:     0/TCP
    Command:
      /metrics-server
      --metric-resolution=30s
      --kubelet-port=10255
      --deprecated-kubelet-completely-insecure=true
      --kubelet-preferred-address-types=InternalIP,Hostname,InternalDNS,ExternalDNS,ExternalIP
```
```
kubectl -n istio-system top pods
```
```
NAME                                      CPU(cores)   MEMORY(bytes)   
grafana-7f4d444dd5-pcxff                  3m           28Mi            
istio-citadel-67d66d65d7-t9srg            3m           15Mi            
istio-galley-dfcc75c46-ljn27              67m          25Mi            
istio-ingressgateway-7497b4c87b-n6pg8     13m          45Mi            
istio-ingressgateway-7497b4c87b-thdcn     7m           45Mi            
istio-pilot-659cbcf679-p9grb              10m          71Mi            
istio-policy-6ccbf85f88-rzh5l             10m          49Mi            
istio-sidecar-injector-79f667994b-ql7p2   19m          9Mi             
istio-telemetry-5db544fc5d-2krjp          60m          71Mi            
istio-tracing-79db5954f-p8glb             1m           42Mi            
kiali-68677d47d7-6qqnp                    1m           9Mi             
prometheus-5977597c75-qq4qq               38m          416Mi
```

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[x] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
ingressgateway hpa should figure out cpu utilization and scale accordingly.

**Steps to reproduce the bug**
install istio on gke using helm template with the following config
```
gateways:
  istio-ingressgateway:
    sds:
      enabled: true
    
    # autoscaling
    autoscaleEnabled: true
    autoscaleMin: 2
    autoscaleMax: 10
    cpu:
      targetAverageUtilization: 70

    # resources
    resources:
      requests:
        cpu: 250m
        memory: 256Mi
      limits:
        cpu: 1000m
        memory: 1Gi

grafana:
  enabled: true

tracing:
  enabled: true

kiali:
  enabled: true
```

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
```
client version: 1.2.2
citadel version: release-1.1-20190731-09-16
galley version: release-1.1-20190731-09-16
ingressgateway version: release-1.1-20190731-09-16
ingressgateway version: release-1.1-20190731-09-16
pilot version: release-1.1-20190731-09-16
policy version: release-1.1-20190731-09-16
sidecar-injector version: release-1.1-20190731-09-16
telemetry version: release-1.1-20190731-09-16
```
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"2018-12-13T19:44:19Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.7-gke.8", GitCommit:"7d3d6f113e933ed1b44b78dff4baf649258415e5", GitTreeState:"clean", BuildDate:"2019-06-19T16:37:16Z", GoVersion:"go1.11.5b4", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
using `kubectl apply` on manifests generated via helm template

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE `v1.13.7-gke.8`

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by aditya-prasad** [comment]
<img width="752" alt="Screenshot 2019-08-01 at 16 35 42" src="https://user-images.githubusercontent.com/2631850/62288667-81e83780-b47a-11e9-87be-d569b84e158f.png">

**Comment by howardjohn** [comment]
@aditya-prasad it says there is no request for CPU? Can you run `kubectl get pod INGRESS -oyaml` and check if that is true or not?

**Comment by aditya-prasad** [comment]
resource request and limit are properly set for `istio-proxy` container but not for `ingress-sds` container
```
kubectl -n istio-system get pod istio-ingressgateway-7497b4c87b-n6pg8 -o yaml
```
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    sidecar.istio.io/inject: "false"
  creationTimestamp: "2019-08-01T10:08:54Z"
  generateName: istio-ingressgateway-7497b4c87b-
  labels:
    app: istio-ingressgateway
    chart: gateways
    heritage: Tiller
    istio: ingressgateway
    pod-template-hash: 7497b4c87b
    release: istio
  name: istio-ingressgateway-7497b4c87b-n6pg8
  namespace: istio-system
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: istio-ingressgateway-7497b4c87b
    uid: 5e1bf684-b444-11e9-8019-42010aa00063
  resourceVersion: "31719205"
  selfLink: /api/v1/namespaces/istio-system/pods/istio-ingressgateway-7497b4c87b-n6pg8
  uid: 5e1e6a8f-b444-11e9-8019-42010aa00063
spec:
  affinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - preference:
          matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - amd64
        weight: 2
      - preference:
          matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - ppc64le
        weight: 2
      - preference:
          matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - s390x
        weight: 2
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - amd64
            - ppc64le
            - s390x
  containers:
  - env:
    - name: ENABLE_WORKLOAD_SDS
      value: "false"
    - name: ENABLE_INGRESS_GATEWAY_SDS
      value: "true"
    - name: INGRESS_GATEWAY_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    image: gcr.io/istio-release/node-agent-k8s:release-1.1-latest-daily
    imagePullPolicy: IfNotPresent
    name: ingress-sds
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/ingress_gateway
      name: ingressgatewaysdsudspath
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: istio-ingressgateway-service-account-token-8vg45
      readOnly: true
  - args:
    - proxy
    - router
    - --domain
    - $(POD_NAMESPACE).svc.cluster.local
    - --log_output_level=default:info
    - --drainDuration
    - 45s
    - --parentShutdownDuration
    - 1m0s
    - --connectTimeout
    - 10s
    - --serviceCluster
    - istio-ingressgateway
    - --zipkinAddress
    - zipkin:9411
    - --proxyAdminPort
    - "15000"
    - --statusPort
    - "15020"
    - --controlPlaneAuthPolicy
    - NONE
    - --discoveryAddress
    - istio-pilot:15010
    env:
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: INSTANCE_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    - name: HOST_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.hostIP
    - name: ISTIO_META_POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: ISTIO_META_CONFIG_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: ISTIO_META_USER_SDS
      value: "true"
    - name: ISTIO_META_ROUTER_MODE
      value: sni-dnat
    image: gcr.io/istio-release/proxyv2:release-1.1-latest-daily
    imagePullPolicy: IfNotPresent
    name: istio-proxy
    ports:
    - containerPort: 15020
      protocol: TCP
    - containerPort: 80
      protocol: TCP
    - containerPort: 443
      protocol: TCP
    - containerPort: 31400
      protocol: TCP
    - containerPort: 15029
      protocol: TCP
    - containerPort: 15030
      protocol: TCP
    - containerPort: 15031
      protocol: TCP
    - containerPort: 15032
      protocol: TCP
    - containerPort: 15443
      protocol: TCP
    - containerPort: 15090
      name: http-envoy-prom
      protocol: TCP
    readinessProbe:
      failureThreshold: 30
      httpGet:
        path: /healthz/ready
        port: 15020
        scheme: HTTP
      initialDelaySeconds: 1
      periodSeconds: 2
      successThreshold: 1
      timeoutSeconds: 1
    resources:
      limits:
        cpu: "1"
        memory: 1Gi
      requests:
        cpu: 250m
        memory: 256Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/ingress_gateway
      name: ingressgatewaysdsudspath
    - mountPath: /etc/certs
      name: istio-certs
      readOnly: true
    - mountPath: /etc/istio/ingressgateway-certs
      name: ingressgateway-certs
      readOnly: true
    - mountPath: /etc/istio/ingressgateway-ca-certs
      name: ingressgateway-ca-certs
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: istio-ingressgateway-service-account-token-8vg45
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: gke-staging-pool-cpu8-mem30-cf648099-53ct
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: istio-ingressgateway-service-account
  serviceAccountName: istio-ingressgateway-service-account
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir: {}
    name: ingressgatewaysdsudspath
  - name: istio-certs
    secret:
      defaultMode: 420
      optional: true
      secretName: istio.istio-ingressgateway-service-account
  - name: ingressgateway-certs
    secret:
      defaultMode: 420
      optional: true
      secretName: istio-ingressgateway-certs
  - name: ingressgateway-ca-certs
    secret:
      defaultMode: 420
      optional: true
      secretName: istio-ingressgateway-ca-certs
  - name: istio-ingressgateway-service-account-token-8vg45
    secret:
      defaultMode: 420
      secretName: istio-ingressgateway-service-account-token-8vg45
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2019-08-01T10:08:54Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2019-08-01T10:08:58Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2019-08-01T10:08:58Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2019-08-01T10:08:54Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://57749559efe4a59a97c591830f928d1e9b2b34026f16898b371c12fbc0637ca2
    image: gcr.io/istio-release/node-agent-k8s:release-1.1-latest-daily
    imageID: docker-pullable://gcr.io/istio-release/node-agent-k8s@sha256:0c228b92775c42a5331c8157602767ead250c308b0efe2f6d44a6e9e210c3419
    lastState: {}
    name: ingress-sds
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: "2019-08-01T10:08:55Z"
  - containerID: docker://df7dfdaf2223aa6940a59c5d050cd62f2ec8e71121e547991d3e3033fecdf38d
    image: gcr.io/istio-release/proxyv2:release-1.1-latest-daily
    imageID: docker-pullable://gcr.io/istio-release/proxyv2@sha256:75a702e6bb42f346ec276203f94f36d730eb8cd7131a9e9c32a47e664c8bd201
    lastState: {}
    name: istio-proxy
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: "2019-08-01T10:08:55Z"
  hostIP: 10.0.1.34
  phase: Running
  podIP: 10.12.2.34
  qosClass: Burstable
  startTime: "2019-08-01T10:08:54Z"
```

**Comment by howardjohn** [comment]
Interesting, are you sure you have the same values set when you installed?

I tried on istio 1.1.7 with your values and the resources show up:

```
$ helm template install/kubernetes/helm/istio --name istio --namespace istio-system -f 15979.yaml | grep ingress-sds -C 10
        chart: gateways
        heritage: Tiller
        release: istio
        app: istio-ingressgateway
        istio: ingressgateway
      annotations:
        sidecar.istio.io/inject: "false"
    spec:
      serviceAccountName: istio-ingressgateway-service-account
      containers:
        - name: ingress-sds
          image: "docker.io/istio/node-agent-k8s:1.1.7"
          imagePullPolicy: IfNotPresent
          resources:
            limits:
              cpu: 2000m
              memory: 1024Mi
            requests:
              cpu: 100m
              memory: 128Mi
```

**Comment by aditya-prasad** [comment]
i was using istio `1.1.0`, seems to be fixed by `1.1.7`. shall upgrade.

https://github.com/istio/istio/blob/1.1.0/install/kubernetes/helm/istio/charts/gateways/templates/deployment.yaml#L62-L85

https://github.com/istio/istio/blob/1.1.7/install/kubernetes/helm/istio/charts/gateways/templates/deployment.yaml#L61-L90

closing this. thanks.
